### PR TITLE
Migrate TestShowMoreCard.js

### LINKED ISSUE #11252: Migrate /components/TestShowMoreCard.js to react-testing-library
Migrate /components/TestShowMoreCard.js to react-testing-library

For QA:

Some refactoring was done to the `ShowMoreCard` component which should not impact behaviour, but should be checked for regressions. This is the card that shows a "Read more" link when the contents of the card are too big. Some places to check this include:
- The Permissions card on the Add-on Detail page
- The Add-on description card on the Add-on Detail page
- The Developer comments card on the Add-on Detail page
- The Release notes card on the Add-on Detail page

It's not necessary to check all of these, but checking a couple would make sense.

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #11252

### CODE REVIEW

**Comment by bobsilverberg** [comment]
Note that the coverage failure is expected and is discussed in a comment in the test suite, and will be addressed via #11409.

File: src/amo/components/ShowMoreCard/index.js [hunk 1]
```diff
@@ -23,20 +23,48 @@ type UIStateType = {|
...
+  setUIState: ($Shape<UIStateType>) => void,
+  uiState: UIStateType,
+|};
+
+export const truncateToMaxHeight: (TruncateToMaxHeightProps) => void = ({
+  contents,
+  maxHeight,
+  setUIState,
+  uiState,
+}: TruncateToMaxHeightProps) => {
```
Comment (eviljeff) [inline src/amo/components/ShowMoreCard/index.js:38]: Why do we have to return all the same parameters that were passed in?  (and so necessitating a weird `TruncateToMaxHeightProps` type.)

**Review by eviljeff — COMMENTED** [review]
Can you add some explanation of what's going on with the rewrite of `truncateToMaxHeight`?  The test changes themselves are pretty straightforward in comparison.

File: src/amo/components/ShowMoreCard/index.js (hunk 1 above)
Comment (bobsilverberg) [reply src/amo/components/ShowMoreCard/index.js]: We don't. It's just that the Flow definitions are a bit hard to follow. I realized we don't need to define the type of the function itself, which removes some of the code above and hopefully makes it clearer. It also means we don't need the extra  `TruncateToMaxHeightProps` type.

File: src/amo/components/ShowMoreCard/index.js [hunk 2]
```diff
@@ -23,20 +23,46 @@ type UIStateType = {|
   readMoreExpanded: boolean,
 |};
 
+export const truncateToMaxHeight = ({
```
Comment (bobsilverberg) [inline src/amo/components/ShowMoreCard/index.js:26]: This function was moved outside of the component's definition so that it can be the default value for `_truncateToMaxHeight` inside the component.

File: src/amo/components/ShowMoreCard/index.js [hunk 3]
```diff
@@ -23,20 +23,46 @@ type UIStateType = {|
...
   contentId: number | string | null,
   children: React.Node,
   className?: string,
   header?: React.Node | string,
   id: string,
+  maxHeight?: number,
 |};
 
 type InternalProps = {|
+  _setUIState?: ($Shape<UIStateType>) => void,
```
Comment (bobsilverberg) [inline src/amo/components/ShowMoreCard/index.js:64]: We also need to be able to mock `setUIState` from a test, which is why this has become a new prop.

File: src/amo/components/ShowMoreCard/index.js [hunk 4]
```diff
@@ -53,19 +79,38 @@ export class ShowMoreCardBase extends React.Component<InternalProps> {
...
+      maxHeight,
+      setUIState,
+      uiState,
+    } = this.props;
     if (!uiState.readMoreExpanded) {
-      this.truncateToMaxHeight(this.contents);
+      _truncateToMaxHeight({
+        contents: this.contents,
+        maxHeight,
+        setUIState: _setUIState || setUIState,
```
Comment (bobsilverberg) [inline src/amo/components/ShowMoreCard/index.js:98]: We cannot use `defaultProps` to set a default value for `setUIState` because the default is a prop that is injected via the `withUIState` HOC, and is therefore not available when `defaultProps` executes. This is why we need this conditional here and also in `componentDidUpdate`.

**Comment by bobsilverberg** [comment]
> Can you add some explanation of what's going on with the rewrite of `truncateToMaxHeight`? The test changes themselves are pretty straightforward in comparison.

I added a few comments that hopefully explain the changes in `ShowMoreCard`. I also changed the Flow definitions for `truncateToMaxHeight` so that they are hopefully easier to follow.

File: src/amo/components/ShowMoreCard/index.js (hunk 2 above)
Comment (eviljeff) [reply src/amo/components/ShowMoreCard/index.js:26]: mmm, but just moving it outside the component changes:
```
truncateToMaxHeight: (contents: HTMLElement | null) => void = (
    contents: HTMLElement | null,
  ) =>
```
to
```
export const truncateToMaxHeight = ({
  contents,
  maxHeight,
  setUIState,
  uiState,
}: {|
  contents: HTMLElement | null,
  maxHeight: number,
  setUIState: ($Shape<UIStateType>) => void,
  uiState: UIStateType,
|}) =>
```
but he functionality hasn't changed?

Comment (bobsilverberg) [reply src/amo/components/ShowMoreCard/index.js:26]: Correct, the functionality has not changed. We now need to accept arguments for `maxHeight`, `setUIState` and `uiState` because previously those were being read from `this.props`, but `this.props` doesn't exist when we move the function outside of the component.

File: src/amo/components/ShowMoreCard/index.js (hunk 3 above)
Comment (eviljeff) [reply src/amo/components/ShowMoreCard/index.js:64]: so `_setUIState` only exists to be able to override `setUIState` in tests?  it's pretty weird to me that's a necessity in our testing setup but okay.

**Review by eviljeff — APPROVED** [review]
still seems like a lot of magic syntax changes to work around Jest or whatever but if that's the way to do it then 🤷
